File: base/sort/quicksort.py
def quicksort(l, start, end):
    if start < end:
        i, j = start, end
        base = l[i]
        while i < j:
            while (i < j) and l[j] >= base:
                j -= 1
            l[i] = l[j]
            while (i < j) and l[i] <= base:
                i += 1
            l[j] = l[i]
        l[i] = base
        quicksort(l, start, i-1)
        quicksort(l, i+1, end)
    return l


myList = [49, 38, 65, 97, 76, 13, 27, 49]
print("Quick Sort: ")
quicksort(myList, 0, len(myList) - 1)
print(myList)

File: java_code/src/base/QuickSort.java
package base;

import java.util.Arrays;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Created by Max on 2018/8/22.
 */
public class QuickSort {

    public static void sort(int[] arr) {
        core(arr, 0, arr.length - 1);
    }

    private static void core(int[] arr, int i, int j) {
        if (i >= j) {
            return;
        }
        int start = i, end = j;
        int baseIndex = ThreadLocalRandom.current().nextInt(i, j);
//        int base = start;

        int baseValue = arr[baseIndex];

        arr[baseIndex] = arr[start];


        while (start < end) {

            while (start < end && arr[end] >= baseValue) {
                end--;
            }
            arr[start] = arr[end];

            while (start < end && arr[start] <= baseValue) {
                start++;
            }
            arr[end] = arr[start];
        }

        arr[start] = baseValue;

        core(arr, i, baseIndex - 1);
        core(arr, baseIndex + 1, j);

    }

    public static void main(String[] args) {
        int[] arr = new int[]{4, 5, 2, 1, 8};
        sort(arr);
        System.out.println(Arrays.toString(arr));
    }

}
